rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,209 @@
+implementations code. It needs to include all code into the binary which was
+referenced from the list. It finds object files (that is, translation units) and
+includes necessary code and may include some unnecessary code too. Unnecessary
+code can be either of:
+* Instantiations of non-requested templated classes.
+* Specializations of non-requested templated classes or class members.
+* Non-templated definitions of non-requested classes.
+
+Non-requested classes refer mostly to backward propagation kind related codes
+and their auxiliary methods. It also refers to ISA-specific instantiations
+which were explicitly disabled by a user. This linker behavior can't be changed
+or influenced externally.
+
+## Implementation Details
+
+Build system will define the following set of macros:
+
+```cpp
+// Primitives list controls
+// Propagation kind controls
+#cmakedefine01 BUILD_TRAINING
+#cmakedefine01 BUILD_INFERENCE
+// Primitives controls
+#cmakedefine01 BUILD_PRIMITIVE_ALL
+#cmakedefine01 BUILD_BATCH_NORMALIZATION
+#cmakedefine01 BUILD_BINARY
+...
+// Primitives ISA controls
+#cmakedefine01 BUILD_PRIMITIVE_ISA_ALL
+#cmakedefine01 BUILD_SSE41
+...
+```
+
+Based on this `BUILD_XXX` macros, the following internal macros would be set:
+
+```cpp
+// src/common/impl_registration.hpp (new header):
+
+#if BUILD_PRIMITIVE_ALL || (BUILD_CONVOLUTION)
+#define REG_CONV_FWD(...) __VA_ARGS__
+#else
+#define REG_CONV_FWD(...)
+#endif
+
+#if BUILD_PRIMITIVE_ALL || (BUILD_CONVOLUTION && BUILD_TRAINING)
+#define REG_CONV_BWD_D(...) __VA_ARGS__
+#define REG_CONV_BWD_W(...) __VA_ARGS__
+#else
+#define REG_CONV_BWD_D(...)
+#define REG_CONV_BWD_W(...)
+#endif
+```
+
+`BWD_D` and `BWD_W` separation applies for convolution only to handle
+deconvolution dependency gracefully. Rest would receive simply `BWD` despite the
+fact they may support data and weights separately.
+
+Finally, these internal macros are applied inside primitive implementation
+lists:
+
+```cpp
+// src/cpu/cpu_convolution_list.cpp:
+const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    // FWD fp
+    {{forward, f32, f32, f32}, {
+        REG_CONV_FWD(REG_IP_FWD(CPU_INSTANCE_X64(ip_convolution_fwd_t)))
+        REG_CONV_FWD(CPU_INSTANCE_X64(brgemm_1x1_convolution_fwd_t<avx512_core>))
+        ...
+        CPU_INSTANCE_AARCH64_ACL(acl_wino_convolution_fwd_t)
+        CPU_INSTANCE_AARCH64(jit_sve_512_dw_convolution_fwd_t)
+        ...
+        REG_CONV_FWD(CPU_INSTANCE(gemm_convolution_fwd_t))
+        REG_CONV_FWD(CPU_INSTANCE(ref_convolution_fwd_t))
+        REG_CONV_FWD(CPU_INSTANCE(ref_fused_convolution_fwd_t))
+        nullptr,
+    }},
+```
+
+Aarch64 implementations wouldn't be a part of change in the first part of
+changes. This example also doesn't show the use of ISA macros, but they would
+be there. It's not decided yet whether this would be a separate macro, or ISA
+name would be incorporated into a single macro as well, like
+`REG_CONV_AVX2_FWD`.
```

Comment:
I guess that for ISA selection, we might need to clarify the behavior first.
For example the technical solution might be different if we need to enable a single impl for 2 ISAs (e.g. AVX and SSE41).